feature-swap-impl/src/main/java/io/novafoundation/nova/feature_swap_impl/presentation/options/SwapOptionsViewModel.kt

```diff
@@ -10,84 +10,87 @@ import io.novafoundation.nova.common.utils.Percent
 import io.novafoundation.nova.feature_swap_api.presentation.state.SwapSettingsStateProvider
 import io.novafoundation.nova.feature_swap_impl.R
-import io.novafoundation.nova.feature_swap_impl.domain.slippage.SlippageRepository
-import io.novafoundation.nova.feature_swap_impl.domain.validation.slippage.SlippageValidationFailure
-import io.novafoundation.nova.feature_swap_impl.domain.validation.slippage.SlippageValidationPayload
+import io.novafoundation.nova.feature_swap_impl.data.assetExchange.assetConversion.AssetConversionExchangeFactory
 import io.novafoundation.nova.feature_swap_impl.presentation.SwapRouter
+import io.novafoundation.nova.feature_swap_impl.presentation.fieldValidation.SlippageFieldValidatorFactory
 import kotlinx.coroutines.async
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.filterNotNull
 import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flatMapLatest
 import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.mapNotNull
 import kotlinx.coroutines.launch
 
 class SwapOptionsViewModel(
     private val swapRouter: SwapRouter,
     private val resourceManager: ResourceManager,
     private val swapSettingsStateProvider: SwapSettingsStateProvider,
-    private val inputValidationMixinFactory: InputValidationMixinFactory<SlippageValidationPayload, SlippageValidationFailure>,
-    private val slippageRepository: SlippageRepository
+    private val assetExchangeFactory: AssetConversionExchangeFactory,
+    private val slippageFieldValidatorFactory: SlippageFieldValidatorFactory
 ) : BaseViewModel() {
 
-    private val defaultSlippage = slippageRepository.getDefaultSlippage()
-    private val slippageTips = slippageRepository.getSlippageTips()
-
     private val swapSettingState = async {
         swapSettingsStateProvider.getSwapSettingsState(viewModelScope)
     }
 
     private val swapSettingsStateFlow = flowOfAll { swapSettingState.await().selectedOption }
         .shareInBackground()
 
+    private val assetExchange = swapSettingsStateFlow.mapNotNull { it.assetIn }
+        .map { assetExchangeFactory.create(it.chainId) }
+        .filterNotNull()
+        .shareInBackground()
+
+    private val slippageConfig = assetExchange.map { it.slippageConfig() }
+        .shareInBackground()
+
     val slippageInput = MutableStateFlow("")
 
-    private val slippageInputValidationStatus = inputValidationMixinFactory.create(slippageInput, ::formatSlippagePayload)
-        .observeStatus()
+    private val slippageFieldValidator = assetExchange.map { slippageFieldValidatorFactory.create(it) }
         .shareInBackground()
 
-    private fun formatSlippagePayload(text: String): SlippageValidationPayload {
-        return SlippageValidationPayload(text.formatToPercent())
-    }
+    private val slippageInputValidationResult = slippageFieldValidator.flatMapLatest { it.observe(slippageInput) }
+        .shareInBackground()
```

Comment:
We should always think about code reusability and code dublication. Given that the very next task is to apply the same logic to multiple fields in the setup swap screen, it is really worth to make field validation easy to reuse from the beginning. Here are some thoughts on how it can potentially be achieved (feel freee to adjust):

``` kotlin
// -------------- views ------------------

interface ValidatableInputField {
  fun showError(error: String)
  
  fun hideError()
}

class SwapInputView : ValidatableInputField 
class SlippageInputView : ValidatableInputField // I dont rember the exact name for this view, idea is to implement interface for every validatrable field
 
// --------- presentation --------------

interface InputValidationMixin {
  val fieldValidationResult: Flow<FieldValidationResult>

  // this factory can be used standalone (for slippage) or inside other mixins like SwapAmountInputMixin
  interface Factory {
    fun create(coroutineScope: CoroutineScope, inputStream: Flow<String>, validator: FieldValidator)
  }
}

class SwapAmountInputMixinFactory(..) {
 fun create(
        ....
        validator: FieldValidator,
    ): SwapAmountInputMixin.Presentation {
        return RealSwapAmountInputMixin(
            ....
            validator: FieldValidator,
        )
    }
}